### ISSUE TITLE
Array declaration spacing sniff

### ISSUE THREAD

**Issue by michnovka** [slevomat/coding-standard#1456]
```php
#[DiscriminatorMap(['cc' => Component\ConcreteComponent::class,
        'cd' => Decorator\ConcreteDecorator::class])]
abstract class Component
```
This looks wrong to me and currently there are no sniffs targeting this.

What do you think about `SlevomatCodingStandard.Attributes.RequireAttributeParamsInline`?
```php
#[DiscriminatorMap(['cc' => Component\ConcreteComponent::class, 'cd' => Decorator\ConcreteDecorator::class])]
 abstract class Component
```
Or we can have a sniff that enforces
```php
#[DiscriminatorMap([
    'cc' => Component\ConcreteComponent::class,
    'cd' => Decorator\ConcreteDecorator::class,
])]
abstract class Component
```

btw, is `SlevomatCodingStandard.Arrays.TrailingArrayComma` applicable here? EDIT: it does apply and does work as expected

**Comment by kukulich** [comment]
I think it should be solved by standard sniffs because it's constructor call.

**Comment by michnovka** [comment]
@kukulich ok, so it looks like what i am missing is actually some Array sniff, like we have `SlevomatCodingStandard.Arrays.MultiLineArrayEndBracketPlacement` I would be looking for `SlevomatCodingStandard.Arrays.MultiLineArrayStartBracketPlacement` or `SlevomatCodingStandard.Arrays.MultiLineArrayFirstItemPlacement`

**Comment by kukulich** [comment]
There are sniff in PHPCS itself, like `Generic.Arrays.ArrayIndent`, that should solve this - I think.

**Comment by michnovka** [comment]
@kukulich I dont know, I have this sniff enabled and it doesnt complain about stuff like

```php
#[ORM\Index(columns: ['phone_prefix',
    'sed',
])]
```

**Comment by kukulich** [comment]
@michnovka The sniff probably does not have the implementation :) However I think it should be it's job.

**Comment by michnovka** [comment]
@kukulich I have found this rule does what I need:

```xml
    <rule ref="WordPress.Arrays.ArrayDeclarationSpacing">
        <exclude name="WordPress.Arrays.ArrayDeclarationSpacing.NoSpaceAfterArrayOpener"/>
        <exclude name="WordPress.Arrays.ArrayDeclarationSpacing.NoSpaceBeforeArrayCloser"/>
    </rule>
```

**Comment by michnovka** [comment]
@kukulich the WP sniffs are very opinionated and change (as the name suggests `WordPress.Arrays.ArrayDeclarationSpacing` means -> use whatever WP standard is now accepted for array declaration spacing. There was case when this very standard [changed](https://github.com/WordPress/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#arrays-forcing-single-item-associative-arrays-to-be-multi-line)

This sniff also does so many things:

>  Enforces WordPress array spacing format.
>  
>  - Check for no space between array keyword and array opener.
>  - Check for no space between the parentheses of an empty array.
>  - Checks for one space after the array opener / before the array closer in single-line arrays.
>  - Checks that associative arrays are multi-line.
>  - Checks that each array item in a multi-line array starts on a new line.
>  - Checks that the array closer in a multi-line array is on a new line.

Slevomat CS on the other hands has an approach to name sniffs properly and have one sniff to one thing only.

Would you be interested in bringing this sniff as multiple sniffs under Slevomat CS?

**Comment by kukulich** [comment]
Yes, it's ok to add such sniff - but check if the sniff does not already exist in PHPCS itself. There are some sniffs like `Squiz.Arrays.ArrayDeclaration` or `Generic.Arrays.ArrayIndent`.

**Comment by michnovka** [comment]
@kukulich seems that `Squiz.Arrays.ArrayDeclaration` has lot of the things provided by WP, but I am still missing
> Checks that each array item in a multi-line array starts on a new line.
So lets add this one, I would call it `SlevomatCodingStandard.Arrays.RequireMultilineStartsOnNewLine`. Thoughts?

**Comment by michnovka** [comment]
Also missing
> Checks that associative arrays are multi-line.

So that might be another sniff Id create.

I will not have time for this until mid November, so lets use that time to finish discussion and specs here. maybe somebody will find the sniffs burried somewhere in the Squiz CS.

**Comment by kukulich** [comment]
We should prepare list of missing features so we can create a propose about the sniffs.

**Comment by michnovka** [comment]
Checks that associative arrays are multi-line. => DOES NOT EXIST
Checks that multiple-value arrays are multi-line. => `Squiz.Arrays.ArrayDeclaration.SingleLineNotAllowed`
Checks that each array item in a multi-line array starts on a new line. => `Squiz.Arrays.ArrayDeclaration.IndexNoNewline`


Check for no space between the parentheses of an empty array => `SlevomatCodingStandard.Arrays.SingleLineArrayWhitespace.SpaceInEmptyArray`
Checks for one space after the array opener / before the array closer in single-line arrays. => `SlevomatCodingStandard.Arrays.SingleLineArrayWhitespace.SpaceAfterArrayOpen`
Checks that the array closer in a multi-line array is on a new line. => `Squiz.Arrays.ArrayDeclaration.CloseBraceNewLine`

----
therefore what we miss is

**Checks that associative arrays are multi-line.**

so that 
```php
$array = ['some_key',];
$arrayAssoc = ['some_key' => 'some_value',];
```
would be invalid, and it would be fixed to
```php
$array = ['some_key',];
$array = [
    'some_key' => 'some_value',
];
```

I propose: `SlevomatCodingStandard.Arrays.RequireAssociativeArraysMultiLine`

**Comment by kukulich** [comment]
> I propose: SlevomatCodingStandard.Arrays.RequireAssociativeArraysMultiLine

Looks good to me. Just to be consistent with other sniff, it should probably be `RequireMultiLineAssociativeArray` (yes, naming is hard)
